File: src/test/java/self/ed/generator/ReducerTest.java
package self.ed.generator;

import static org.junit.jupiter.api.Assertions.assertEquals;

import static self.ed.util.FileUtils.readClasspathFile;
import static self.ed.util.SudokuUtils.countOpen;
import static self.ed.util.SudokuUtils.fromString2D;
import static self.ed.util.SudokuUtils.toString2D;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

public class ReducerTest {

  private Reducer reducer;

  @BeforeEach
  void setUp() {
    reducer = new Reducer();
  }

  @Test
  void reduce_alreadyMinimal() {
    int[][] input = fromString2D(readClasspathFile("input-21.txt"));
    int[][] output = reducer.reduce(input);
    assertEquals(toString2D(input), toString2D(output));
  }

  @Test
  void reduce_minusOne() {
    int[][] input = fromString2D(readClasspathFile("input-22.txt"));
    int[][] output = reducer.reduce(input);
    assertEquals(countOpen(input) - 1, countOpen(output));
  }

  @Test
  void reduce_minusTwo() {
    int[][] input = fromString2D(readClasspathFile("input-24.txt"));
    int[][] output = reducer.reduce(input);
    assertEquals(countOpen(input) - 2, countOpen(output));
  }
}

File: cpp-sudoku/test/generator/ReducerTest.cpp
#include "../../src/test/TestFramework.hpp"
#include "../../src/generator/Reducer.hpp"
#include "../../src/util/SudokuUtils.hpp"

using namespace just::demo::generator;
using namespace just::demo::util;
using namespace test;

void testReduceAlreadyMinimal() {
    std::string inputData = R"(
. . . 8 7 . . . .
. . . . . . . 5 9
3 . . 1 . . . . .
. . . . 4 . 2 1 .
5 8 . 7 . . . . .
6 . . . . . . . 4
. 2 . . 5 . 8 . .
. . . . 3 4 . 9 .
. . . . . . . 3 .
)";

    auto input = SudokuUtils::fromString2D(inputData);
    Reducer reducer;
    auto output = reducer.reduce(input);
    
    TestFramework::assertEqual(SudokuUtils::toString2D(input), SudokuUtils::toString2D(output), "testReduceAlreadyMinimal");
}

void testReduceMinusOne() {
    std::string inputData = R"(
. 5 4 . 1 3 . . .
6 . . . . . . . 2
. . . . . 5 . 7 .
. 8 . 2 . . 7 . .
4 . . . . . . . .
. . 6 . . . 9 . 1
. . 2 . . . 6 . .
. . 1 . 4 . . 8 .
. . . . . 8 5 . .
)";

    auto input = SudokuUtils::fromString2D(inputData);
    Reducer reducer;
    auto output = reducer.reduce(input);
    
    TestFramework::assertEqual(SudokuUtils::countOpen(input) - 1, SudokuUtils::countOpen(output), "testReduceMinusOne");
}

void testReduceMinusTwo() {
    std::string inputData = R"(
2 . 9 . . . 3 . .
. . . . 2 4 6 . .
1 . . . 7 . . 5 .
. 1 . . . . . . .
3 4 . . . . . 7 .
. 5 . . . . 9 4 8
4 . . . . 8 . . .
. . . 6 . . . . .
. 2 . 3 5 . . 6 .
)";

    auto input = SudokuUtils::fromString2D(inputData);
    Reducer reducer;
    auto output = reducer.reduce(input);
    
    TestFramework::assertEqual(SudokuUtils::countOpen(input) - 2, SudokuUtils::countOpen(output), "testReduceMinusTwo");
}

int main() {
    std::cout << "Running Reducer Tests..." << std::endl;
    testReduceAlreadyMinimal();
    testReduceMinusOne();
    testReduceMinusTwo();
    std::cout << "Reducer Tests completed." << std::endl;
    return 0;
}
